core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,555 @@
+package org.infinispan.util.concurrent.locks.impl;
+
+import org.infinispan.util.TimeService;
+import org.infinispan.util.concurrent.TimeoutException;
+import org.infinispan.util.concurrent.locks.DeadlockChecker;
+import org.infinispan.util.concurrent.locks.DeadlockDetectedException;
+import org.infinispan.util.concurrent.locks.ExtendedLockPromise;
+import org.infinispan.util.concurrent.locks.LockListener;
+import org.infinispan.util.concurrent.locks.LockState;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import static org.infinispan.util.concurrent.CompletableFutures.await;
+
+/**
+ * A special lock for Infinispan cache.
+ * <p>
+ * The main different with the traditional {@link java.util.concurrent.locks.Lock} is allowing to use any object as lock
+ * owner. It is possible to use a {@link Thread} as lock owner that makes similar to {@link
+ * java.util.concurrent.locks.Lock}.
+ * <p>
+ * In addition, it has an asynchronous interface. {@link #acquire(Object, long, TimeUnit)}  will not acquire the lock
+ * immediately (except if it is free) but will return a {@link ExtendedLockPromise}. This promise allow to test if the
+ * lock is acquired asynchronously and cancel the lock acquisition, without any blocking.
+ *
+ * @author Pedro Ruivo
+ * @since 8.0
+ */
+public class InfinispanLock {
+
+   private static final Log log = LogFactory.getLog(InfinispanLock.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private final Queue<LockPlaceHolder> pendingRequest;
+   private final ConcurrentMap<Object, LockPlaceHolder> lockOwners;
+   private final AtomicReferenceFieldUpdater<InfinispanLock, LockPlaceHolder> fieldUpdater;
+   private final Runnable releaseRunnable;
+   private volatile TimeService timeService;
+   private volatile LockPlaceHolder current;
```

Comment:
Oops, `AbstractQueuedSynchronizer` inherits an `exclusiveOwnerThread` field from `AbstractOwnableSynchronizer`... perhaps in their case it's not worth creating the queue unless there is actual contention, but you create the queue all the time anyway.